doc/static-route/SONiC_static_route_bfd_hld.md

```diff
@@ -0,0 +1,300 @@
+
+<span style="color:blue"> <br>
+*Note: This document describes how to create local BFD session for a static route based on static route configuration. Creating BFD session in its peer system is out of the scope of this document.*<br />  
+</span>
+
+
+# Static Route BFD Design
+## Functional Requirements
+
+1. Create BFD session for each static route nexthop based on static route configuration <br>
+2. Delete BFD session if the nexthop is not needed (triggered by removing static route from configration)<br>
+3. Install, update, removing static route (i.e., update nexthop, remove static route if no nexthop) based on BFD session state change.<br>
+4. When the application (static route BFD) restarts, recover static routes and BFD session states from redis database (config_db, appl_db and state_db) without impacting any existing installed static routes and BFD sessions. <br  />  
+
+## System Overview with Static Route BFD
+A new component, StaticRouteBfd is introduced to support static route BFD. <br>
+
+In the following diagram, StaticRouteBfd monitors config_db STATIC_ROUTE_TABLE. If there is a static route configured with bfd, StaticRouteBfd creates BFD session by updating appl_db BFD_SESSION_TABLE. StaticRouteBfd also monitors state db for BFD session state change. Based on BFD session state, StaticRouteBfd updates appl_db STATIC_ROUTE_TABLE to add/delete static route or update its nexthop list.<br>
+
+To work with existing bgpcfgd StaticRouteMgr, an optional field "bfd" is introduced in STATIC_ROUTE_TABLE. When the "bfd" field present and is "true" in config db, StaticRouteMgr ignores this static route and StaticRouteBfd handles it. StaticRouteBfd writes this static route to appl_db based on BFD session state for this static route nexthop. <br>
+
+To avoid StaticRouteTimer deleting StaticRouteBfd created static route entry in appl_db, a new field "expiry" is introduced in appl_db STATIC_ROUTE_TABLE schema. StaticRouteBfd sets "expiry"="false" when it writes a static route to appl_db STATIC_ROUTE_TABLE. When StaticRouteTimer see "expiry"="false", it ignores this static route entry. <br>
+
+
+<img src="static_rt_bfd_overview.png" width="500">
+<br />
+
+## BFD session local IP address
+StaticRouteBfd create BFD session for each nexthop inside a BFD enabled static route. The BFD peer IP is the nexthop, the BFD local address (source IP address of the BFD packet) is the lookup result using the following method:<br />
+
+1. find the interface name for the nexthop
+2. lookup the config_db, find the IP address of that interface
+3. choose the IP address which has same type of nexthop (IPv4 or IPv6)
+4. If no IP address found, use local loopback IP address, and log a warning message for that. <br /><br />
```

Comment:
what if there are more than one local loopback address?

Reply:
currently it looks up 2 loopback ip addresses:
LOOPBACK0_NAME = "Loopback0"
LOOPBACK4096_NAME = "Loopback4096"
(see the code PR, the link is in this PR description)

Reply:
I'm okay as long as this is assumed by bgpcfgd and works for all cases.